library/core/src/mem/mod.rs

```diff
@@ -772,6 +773,128 @@ pub(crate) const fn swap_simple<T>(x: &mut T, y: &mut T) {
     }
 }
 
+// This version swaps 2 values by chunks of usize and smaller
+// using unaligned reads and writes because they are cheap
+// on modern x86_64 processors for at least 10 years now (at 2022-07-04).
+// https://lemire.me/blog/2012/05/31/data-alignment-for-speed-myth-or-reality/
+// It generates less instructions and memory accesses as well: https://godbolt.org/z/KqGjzPqr1
+// Feel free to add another targets as well if they have fast unaligned accesses.
+//
+// This have to be done by backend optimizer but it currently fails to do so.
+#[cfg(target_arch = "x86_64")]
+#[rustc_const_unstable(feature = "const_swap", issue = "83163")]
+#[inline]
+const fn swap_chunked<T: Sized>(x: &mut T, y: &mut T) {
+    // Rust currently doesn't support constant declarations
+    // which depend from generic parameters.
+    // We workaround this using associated constants from trait.
+    //
+    // Those constants are calculated before emitting code for backend
+    // so
+    // 1. code is much easier to optimize for backend: first DCE removes everything unneccessary;
+    // 2. it would run fast in debug builds too.
+    trait ConstifyOffsets: Sized {
+        const OFF_U64: usize;
+        const COUNT_U64: usize;
+        const OFF_U32: usize;
+        const COUNT_U32: usize;
+        const OFF_U16: usize;
+        const COUNT_U16: usize;
+        const OFF_U8: usize;
+        const COUNT_U8: usize;
+
+        // This would cause compilation error when accessed
+        // if we calculated offsets improperly.
+        const STATIC_TEST: usize = 0usize
+            - (size_of::<Self>() - Self::OFF_U8 - Self::COUNT_U8)
+            - (size_of::<Self>()
+                - (Self::COUNT_U64 * size_of::<u64>()
+                    + Self::COUNT_U32 * size_of::<u32>()
+                    + Self::COUNT_U16 * size_of::<u16>()
+                    + Self::COUNT_U8 * size_of::<u8>()));
+    }
+
+    macro_rules! gen_count {
+        ($ChunkTy:ty, $offset:expr) => {
+            if size_of::<$ChunkTy>() <= size_of::<usize>() {
+                (size_of::<Self>() - $offset) / size_of::<$ChunkTy>()
+            } else {
+                // Don't try to split by chunks smaller than pointer
+                0
+            }
+        };
+    }
+
+    impl<T: Sized> ConstifyOffsets for T {
+        const OFF_U64: usize = 0;
+        const COUNT_U64: usize = gen_count!(u64, Self::OFF_U64);
+
+        const OFF_U32: usize = Self::OFF_U64 + Self::COUNT_U64 * size_of::<u64>();
+        const COUNT_U32: usize = gen_count!(u32, Self::OFF_U32);
+
+        const OFF_U16: usize = Self::OFF_U32 + Self::COUNT_U32 * size_of::<u32>();
+        const COUNT_U16: usize = gen_count!(u16, Self::OFF_U16);
+
+        const OFF_U8: usize = Self::OFF_U16 + Self::COUNT_U16 * size_of::<u16>();
+        const COUNT_U8: usize = gen_count!(u8, Self::OFF_U8);
+    }
+
+    // Trigger static assertion.
+    let _ = <T as ConstifyOffsets>::STATIC_TEST;
```

Comment:
Rather than playing games with overflow, this can just be a `const { assert!(...); }` block.

Reply:
Yes, this is better. Updated.